src/include/server/mir/scene/surface_state_stack.h

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright © 2021 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: William Wold <william.wold@canonical.com>
+ */
+
+#ifndef MIR_SCENE_SURFACE_STATE_STACK_
+#define MIR_SCENE_SURFACE_STATE_STACK_
+
+#include "mir_toolkit/common.h"
+
+namespace mir
+{
+namespace scene
+{
+/// Represents a surface's current state, as well as the states it will fall back to as states are removed. For example
+/// A minimized surface may also be fullscreen (which will become relevant when it's unminimized) but not maximized
+/// (which will become relevant when it's unfullscreened). This class keeps track of all the active and inactive states.
+/// States have this order of precedence (earlier ones will be the active state if present):
+/// - hidden, minimized, fullscreen, attached, maximized/vertmaximized/horizmaximized, restored
+/// This struct can NOT be safely accessed by multiple threads at once.
+struct SurfaceStateStack
```

Comment:
I think "stack" sets algorithmic expectations that are not what this does. In particular, there are no "push/pop" operations.

How does `SurfaceStateTracker` sound?

Reply:
Understandable, as I worked on this I was noticing the API drifted further and further from the original vision of a stack. `SurfaceStateTracker` is as good as anything I can think of.